File: modules/chemical/src/util/transform.cpp
#include "transform.hpp"

#include <glm/gtc/matrix_transform.hpp>

namespace Chemical {

  glm::mat4 Transform::ToMatrix() const {
    glm::mat4 m = glm::mat4(1.0f);

    m = glm::translate(m, position);
    m = glm::rotate(m, glm::radians(rotation.z), glm::vec3(0, 0, 1));
    m = glm::rotate(m, glm::radians(rotation.y), glm::vec3(0, 1, 0));
    m = glm::rotate(m, glm::radians(rotation.x), glm::vec3(1, 0, 0));
    m = glm::scale(m, scale);

    return m;
  }
} // namespace Chemical

File: chemical-api/src/scene/types/transform.rs
use glam::{Mat4, Quat, Vec3};
#[derive(Copy, Clone, Debug)]
pub struct Transform {
    pub position: Vec3,
    pub scale: Vec3,
    pub orientation: Quat,
}
impl Transform {
    pub fn to_matrix(&self) -> Mat4 {
        Mat4::from_scale_rotation_translation(self.scale, self.orientation, self.position)
    }
}

impl Default for Transform {
    fn default() -> Transform {
        Transform {
            position: (0.0, 0.0, 0.0).into(),
            scale: (1.0, 1.0, 1.0).into(),
            orientation: Quat::from_xyzw(0.0, 0.0, 0.0, -1.0),
        }
    }
}
/*
impl From<&Transform> for Mat4 {
    fn from(v: &Transform) -> Self {
        Mat4::from_scale_rotation_translation(v.scale, v.orientation, v.position)
    }
}*/
/*impl From<&Transform> for [[f32; 4]; 4] {
    fn from(v: &Transform) -> Self {
        Mat4::from_scale_rotation_translation(v.scale, v.orientation, v.position).to_cols_array_2d()
    }
}
*/
